app/Command/Compile.hs

```diff
@@ -30,26 +32,54 @@ data PSCMakeOptions = PSCMakeOptions
   , pscmOpts         :: P.Options
   , pscmUsePrefix    :: Bool
   , pscmJSONErrors   :: Bool
+  , pscmStrict       :: Bool
   }
 
--- | Arguments: verbose, use JSON, warnings, errors
-printWarningsAndErrors :: Bool -> Bool -> P.MultipleErrors -> Either P.MultipleErrors a -> IO ()
-printWarningsAndErrors verbose False warnings errors = do
+-- | Arguments: verbose, use JSON, strict, warnings, errors
+printWarningsAndErrors :: Bool -> Bool -> Bool -> P.MultipleErrors -> Either P.MultipleErrors a -> IO ()
+printWarningsAndErrors verbose False strict warnings errors = do
   pwd <- getCurrentDirectory
   cc <- bool Nothing (Just P.defaultCodeColor) <$> ANSI.hSupportsANSI stdout
   let ppeOpts = P.defaultPPEOptions { P.ppeCodeColor = cc, P.ppeFull = verbose, P.ppeRelativeDirectory = pwd }
-  when (P.nonEmpty warnings) $
-    putStrLn (P.prettyPrintMultipleWarnings ppeOpts warnings)
-  case errors of
+      (warnings', errors') = promoteSrcWarnings strict warnings errors
+  when (P.nonEmpty warnings') $
+    putStrLn (P.prettyPrintMultipleWarnings ppeOpts warnings')
+  case errors' of
     Left errs -> do
       putStrLn (P.prettyPrintMultipleErrors ppeOpts errs)
       exitFailure
     Right _ -> return ()
-printWarningsAndErrors verbose True warnings errors = do
+printWarningsAndErrors verbose True strict warnings errors = do
+  let (warnings', errors') = promoteSrcWarnings strict warnings errors
   putStrLn . LBU8.toString . A.encode $
-    JSONResult (toJSONErrors verbose P.Warning warnings)
-               (either (toJSONErrors verbose P.Error) (const []) errors)
-  either (const exitFailure) (const (return ())) errors
+    JSONResult (toJSONErrors verbose P.Warning warnings')
+               (either (toJSONErrors verbose P.Error) (const []) errors')
+  either (const exitFailure) (const (return ())) errors'
+
+-- |
+-- Moves all warnings emitted from files in the `src/` directory
+-- into the errors.
+promoteSrcWarnings :: Bool -> P.MultipleErrors -> Either P.MultipleErrors a -> (P.MultipleErrors, Either P.MultipleErrors a)
+promoteSrcWarnings False w e = (w, e)
+promoteSrcWarnings True w e = (P.MultipleErrors nonSrcWarnings, srcWarningsAndErrors)
+  where
+  srcWarningsAndErrors = case e of
+    Left errors -> Left $ P.MultipleErrors $ srcWarnings <> P.runMultipleErrors errors
+    r@(Right _)
+      | null srcWarnings -> r
+      | otherwise -> Left $ P.MultipleErrors srcWarnings
+
+  (srcWarnings, nonSrcWarnings) =
+    partition isSrcWarning $ P.runMultipleErrors w
+
+  isSrcWarning :: P.ErrorMessage -> Bool
+  isSrcWarning warning = fromSourceDir $ fromMaybe "" $ do
+    spans <- P.errorSpan warning
+    pure $ P.spanName $ NEL.head spans
+    where
+    -- account for both POSIX and Windows path separators
+    fromSourceDir fileName =
+      isPrefixOf "src/" fileName || isPrefixOf "src\\" fileName
```

Comment:
I'm not sure we want to hard code this to `src/` in case someone wants to promote `test/` warnings as well.

Reply:
Yes, I think this needs to be configurable because I'm sure people are putting source files in places other than `src` in apps. I'd suggest using the same strategy as psa does - having a list of directory names that indicate that a source file should not count as a local project source file if one of its parent directories matches an entry in the list. So I think we need an `--is-lib` option as well.

Reply:
> Yes, I think this needs to be configurable because I'm sure people are putting source files in places other than src in apps.

Note this is why the predicate is for positively matching library sources. Anything that's not a library source doesn't have to be specified.

Reply:
I think we should still make it configurable, but I just mean to say that I'd wager it's extremely uncommon to have custom library sources.

Reply:
>  I just mean to say that I'd wager it's extremely uncommon to have custom library sources.

That might change when the registry gets implemented and we also implement targets.

Reply:
Latest commit adds the `--is-lib` using a comma-separated value, with the default value being `.spago,bower_components`.

Reply:
Is making this option to be used like `--strict=src/**/*.purs` (which could be passed multiple times for multiple locations) would not work out (without `--is-lib`)?